wnfs/src/private/directory.rs

```diff
@@ -616,6 +616,94 @@ impl PrivateDirectory {
         }
     }
 
+    /// Opens a mutable reference to the specified file.
+    /// If the file is missing, it initializes an empty file and give a mut reference to that.
+    /// If the file already exists, it will copy it to the next revision, update the edit time, and give a mut reference to that.
+    /// # Examples
+    /// ```
+    /// use std::rc::Rc;
+    /// use chrono::Utc;
+    /// use rand::thread_rng;
+    /// use wnfs::{
+    ///    private::{PrivateForest, PrivateRef, PrivateDirectory},
+    ///    common::{BlockStore, MemoryBlockStore},
+    ///    namefilter::Namefilter,
+    /// };
+    /// #[async_std::main]
+    /// async fn main() {
+    ///    let store = &mut MemoryBlockStore::default();
+    ///    let rng = &mut thread_rng();
+    ///    let forest = &mut Rc::new(PrivateForest::new());
+    ///    let root_dir = &mut Rc::new(PrivateDirectory::new(
+    ///         Namefilter::default(),
+    ///         Utc::now(),
+    ///         rng,
+    ///    ));
+    ///    let content = b"print('hello world')";
+    ///    root_dir
+    ///        .write(
+    ///            &["code".into(), "hello.py".into()],
+    ///            true,
+    ///            Utc::now(),
+    ///            content.to_vec(),
+    ///            forest,
+    ///            store,
+    ///            rng
+    ///        )
+    ///        .await
+    ///        .unwrap();
+    ///    let mut file = {
+    ///        root_dir
+    ///            .open_file_mut(&["code".into(), "hello.py".into()], true, Utc::now(), forest, store, rng)
+    ///            .await
+    ///            .unwrap()
+    ///    };
+    ///    file.set_content(
+    ///        Utc::now(),
+    ///        &b"print('hello world 2')"[..],
+    ///        forest,
+    ///        store,
+    ///        rng,
+    ///    );
+    ///    let result = root_dir
+    ///        .read(&["code".into(), "hello.py".into()], true, forest, store)
+    ///        .await
+    ///        .unwrap();
+    ///    assert_eq!(&result, b"print('hello world 2')");
```

Comment:
i think an RwLock-based one still won't satisfy the borrow checker because there are multiple consecutive functions calling wanting `store: &'a mut impl BlockStore` as an argument.

Reply:
> i think an RwLock-based one still won't satisfy the borrow checker because there are multiple consecutive functions calling wanting `store: &'a mut impl BlockStore` as an argument.

Yeah that's true, but a `RwLock`-based block store should enable a cheap `.clone()` implementation. Then it should be possible to provide two `&mut impl BlockStore`s.

That said, I'm currently wrapping my head around "what if it's just `impl BlockStore`" and whether it's implemented on `&mut MyBlockStore` or `&MyBlockStore` should be (literally) an implementation detail :thinking: 

We may be able to use rust better here.